src/wp-includes/functions.php

```diff
@@ -7984,6 +7984,20 @@ function wp_get_direct_php_update_url() {
 	return $direct_update_url;
 }
 
+/**
+ * Replaces periods with encoded entities in URLs.
+ *
+ * If the URL ends with a period, replace it with %2E to avoid misinterpretation from some email clients.
+ *
+ * @since 5.9.0
+ *
+ * @param string $url A provided URL.
+ * @return string The new encoded URL.
+ */
+function wp_half_baked_url_encode( $url ) {
+	return preg_replace( '/\.$/', '%2E', rawurlencode( $url ) );
+}
+
```

Comment:
```suggestion
/**
 * Replaces periods with encoded entities in URLs.
 *
 * If the URL ends with a period, replace it with %2E to avoid misinterpretation from some email clients.
 *
 * @since 5.9.0
 *
 * @param string $url The URL to encode.
 * @return string The new encoded URL.
 */
function wp_url_encode_ending_period( $url ) {
	if ( ! is_string( $url ) ) {
		return '';
	}

	$url = trim( $url );
	if ( '' === $url ) {
		return '';
	}

	$url = rawurlencode( $url );

	// If the URL ends in a period, replace it with %2E.
	if ( str_ends_with( $url, '.' ) ) {
		$url  = rtrim( $url, '.' );
		$url .= '%2E';
	}

	return $url;
}

```

Some changes:

* Renames to tell what it does when invoked, i.e. it encodes an ending period in a URL.
* Guards to ensure a string is given as a URL to avoid errors when passing to PHP native functions.
* Trims any surrounding spaces.
* Uses `str_ends_with()` and `rtrim() instead of `prep_replace()`. Why? More performant. `str_ends_with()` polyfill was added to Core during 5.9. It's now available for use. Here's the benchmark for it https://3v4l.org/WdWih#v8.0.12